azurelinuxagent/common/cgroups.py

```diff
@@ -0,0 +1,698 @@
+        return results
+
+
+class CGroups(object):
+    """
+    This class represents the cgroup folders for the agent or an extension. This is a pretty lightweight object
+    without much state worth preserving; it's not unreasonable to create one just when you need it.
+    """
+    # whether cgroup support is enabled
+    _enabled = True
+    _hierarchies = CGroupsTelemetry.hierarchies()
+    _use_systemd = None     # Tri-state: None (i.e. "unknown"), True, False
+
+    @staticmethod
+    def for_extension(name):
+        return CGroups(name, CGroups.construct_custom_path_for_hierarchy)
+
+    @staticmethod
+    def for_systemd_service(name):
+        return CGroups(name.lower(), CGroups.construct_systemd_path_for_hierarchy)
+
+    def __init__(self, name, path_maker):
+        """
+        Construct CGroups object. Create appropriately-named directory for each hierarchy of interest.
+
+        :param str name: Name for the cgroup (usually the full name of the extension)
+        :param path_maker: Function which constructs the root path for a given hierarchy where this cgroup lives
+        """
+        if name == "":
+            self.name = "Agents+Extensions"
+            self.is_wrapper_cgroup = True
+        else:
+            self.name = name
+            self.is_wrapper_cgroup = False
+
+        self.cgroups = {}
+
+        if not self.enabled():
+            return
+
+        system_hierarchies = os.listdir(BASE_CGROUPS)
+        for hierarchy in CGroups._hierarchies:
+            if hierarchy not in system_hierarchies:
+                raise CGroupsException("Hierarchy {0} is not mounted".format(hierarchy))
+
+            if self.is_wrapper_cgroup:
+                cgroup_path = path_maker(hierarchy)
+            else:
+                cgroup_path = os.path.join(path_maker(hierarchy), self.name)
+            if not os.path.exists(cgroup_path):
+                logger.info("Creating cgroup directory {0}".format(cgroup_path))
+                CGroups._try_mkdir(cgroup_path)
+            self.cgroups[hierarchy] = cgroup_path
+
+    @staticmethod
+    def is_systemd_manager():
+        """
+        Determine if systemd is managing system services. Many extensions are structured as a set of services,
+        including the agent itself; systemd expects those services to remain in the cgroups in which it placed them.
+        If this process (presumed to be the agent) is in a cgroup that looks like one created by systemd, we can
+        assume systemd is in use.
+
+        :return: True if systemd is managing system services
+        :rtype: Bool
+        """
+        if CGroups._use_systemd is None:
+            hierarchy = HIERARCHIES[0]
+            path = CGroups.get_my_cgroup_folder(hierarchy)
+            CGroups._use_systemd = path.startswith(CGroups.construct_systemd_path_for_hierarchy(hierarchy))
+        return CGroups._use_systemd
+
+    @staticmethod
+    def _try_mkdir(path):
```

Comment:
There is a method to create a directory in fileutil.py. This code handles more cases.  I leave it to you to decide if they should be combined or not.

Reply:
They're distinct. The version here changes exception types to make them more focused to my purpose (making cgroup directories).